### PR TITLE
Fixed #30591 -- Fixed recreation of foreign key constraints on MySQL when altering type of referenced unique field.

### PR DESCRIPTION
**Pull request by uadnan**
Unittests provided by @felixxm.
Ref: https://code.djangoproject.com/ticket/30591

### CODE REVIEW

**Comment by uadnan** [comment]
@felixxm any idea what is the version of MySQL against which unit tests are being executed? I tried to run tests against `5.6.45` and that passes locally. Wondering if that is related to some specific version of MySQL?

**Comment by felixxm** [comment]
Failure is on MariaDB 10.2.24. You can check database versions used by CI on [wiki](https://code.djangoproject.com/wiki/Jenkins).

**Comment by uadnan** [comment]
Thanks @felixxm

**Comment by uadnan** [comment]
It seems like the issue is fixed and the failed test is because of ```django.core.exceptions.ImproperlyConfigured: Error loading psycopg2 module: No module named 'psycopg2'```

Is there a way to trigger the test again?

File: docs/releases/3.0.txt [hunk 1]
```diff
@@ -262,7 +262,8 @@ Management Commands
 Migrations
 ~~~~~~~~~~
 
-* ...
+* Fixed MySQL foreign key constraint error when altering type of unique field
+  referenced by Foreign Key (:ticket:`30591`).
```
Comment (felixxm) [inline docs/releases/3.0.txt]: Release notes are not necessary.

File: django/db/backends/mysql/introspection.py [hunk 2]
```diff
@@ -217,14 +217,18 @@ def get_constraints(self, cursor, table_name):
...
+                        'columns': OrderedSet(),
+                        'primary_key': False,
+                        'unique': False,
+                        'index': False,
+                        'check': True,
+                        'foreign_key': None,
+                    }
+                constraints[constraint]['check'] = True
+                for column in self._parse_constraint_columns(check_clause, columns):
+                    constraints[constraint]['columns'].add(column)
```
Comment (felixxm) [inline django/db/backends/mysql/introspection.py]: This changes looks unrelated and all tests pass without them.

File: django/db/backends/base/schema.py [hunk 3]
```diff
@@ -629,7 +628,7 @@ def _alter_field(self, model, old_field, new_field, old_type, new_type,
         if old_db_params['check'] != new_db_params['check'] and old_db_params['check']:
             meta_constraint_names = {constraint.name for constraint in model._meta.constraints}
             constraint_names = self._constraint_names(
-                model, [old_field.column], check=True,
+                model, [old_field.column], check=True, unique=False,
```
Comment (felixxm) [inline django/db/backends/base/schema.py]: Check constraints don't seem to be related with this issue, moreover `unique` is `False` for all check constraints. Please revert this unrelated change.

File: django/db/backends/base/schema.py [hunk 4]
```diff
@@ -31,10 +31,9 @@ def _is_relevant_relation(relation, altered_field):
 def _related_non_m2m_objects(old_field, new_field):
     # Filter out m2m objects from reverse relations.
     # Return (old_relation, new_relation) tuples.
-    return zip(
-        (obj for obj in old_field.model._meta.related_objects if _is_relevant_relation(obj, old_field)),
-        (obj for obj in new_field.model._meta.related_objects if _is_relevant_relation(obj, new_field))
-    )
+    for rel_name, old_rel in old_field.model._meta.fields_map.items():
+        if _is_relevant_relation(old_rel, old_field):
+            yield old_rel, new_field.model._meta.fields_map[rel_name]
```
Comment (felixxm) [inline django/db/backends/base/schema.py]: This looks unrelated.

File: django/db/backends/mysql/introspection.py (hunk 2 above)
Comment (uadnan) [reply django/db/backends/mysql/introspection.py]: @felixxm, no doubt tests pass for MySQL, but that fails for MariaDB. On debugging I found that `UNIQUE` constraint was removed on `code` column that afterward leads to Foreign key constraint. 

This change is done purposefully. `UNIQUE` constraint is overridden by `check` constraint. This change will merge both constraints thus afterward I can filter `UNIQUE` constraints.

File: django/db/backends/base/schema.py (hunk 3 above)
Comment (uadnan) [reply django/db/backends/base/schema.py]: @felixxm I explained the reason for this change above. That change was necessary to fix tests failing for MariaDB (there might be some other way but I thought that is the case)

File: django/db/backends/mysql/introspection.py (hunk 2 above)
Comment (felixxm) [reply django/db/backends/mysql/introspection.py]: :thinking: All tests pass on MariaDB for me.

File: django/db/backends/base/schema.py (hunk 4 above)
Comment (uadnan) [reply django/db/backends/base/schema.py]: @felixxm without this in-place `test_alter_field_reloads_state_on_fk_with_to_field_related_name_target_type_change (migrations.test_operations.OperationTests)` fails on MariaDB

File: django/db/backends/mysql/introspection.py (hunk 2 above)
Comment (felixxm) [reply django/db/backends/mysql/introspection.py]: If it is possible that `unique` and `check` constraints return by different queries have the same name on `MariaDB ` we should fix this separately. I can dig into it.

File: django/db/backends/base/schema.py (hunk 3 above)
Comment (uadnan) [reply django/db/backends/base/schema.py]: Just for the record: Without this in-place `test_alter_field_reloads_state_on_fk_with_to_field_target_type_change (migrations.test_operations.OperationTests)` fails for MariaDB

```
MySQLdb._exceptions.OperationalError: (1005, 'Can\'t create table `test_django`.`alter_alter_field_reloads_state_on_fk_with_to_field_target_t7bb8` (errno: 150 "Foreign key constraint is incorrectly formed")')
```

File: django/db/backends/mysql/introspection.py (hunk 2 above)
Comment (uadnan) [reply django/db/backends/mysql/introspection.py]: @felixxm but that was failing on Jenkins too. (I'm sorry I can't find the history of old failed test reports for reference).

Comment (uadnan) [reply django/db/backends/mysql/introspection.py]: Yea I'm confident that was the case. Anyway keep me updated about your findings.

Comment (felixxm) [reply django/db/backends/mysql/introspection.py]: I added PR #11704.

Comment (felixxm) [reply django/db/backends/mysql/introspection.py]: Merged.

Comment (uadnan) [reply django/db/backends/mysql/introspection.py]: Thanks @felixxm

**Comment by felixxm** [comment]
@uadnan Thanks for updates :+1: I pushed edits.

File: django/db/backends/base/schema.py (hunk 4 above)
Comment (uadnan) [reply django/db/backends/base/schema.py]: Sounds good @felixxm
